lib/common/strings.c

```diff
@@ -148,6 +150,129 @@ crm_parse_int(const char *text, const char *default_text)
     return (int) result;
 }
 
+/*!
+ * \internal
+ * \brief Scan a double-precision floating-point value from a string
+ *
+ * \param[in]  text      String to scan
+ * \param[out] result    If not NULL, where to store scanned value
+ * \param[out] end_text  If not NULL, where to store pointer to just
+ *                       after value
+ *
+ * \return Standard Pacemaker return code (also set errno on error)
+ */
+static int
+pcmk__scan_double(const char *text, double *result, char **end_text)
+{
+    double local_result = -1.0;
+    char *local_end_text = NULL;
+    int rc = pcmk_rc_ok;
+
+    errno = 0;
+    if (text != NULL) {
+        local_result = strtod(text, &local_end_text);
+
+        if (errno == ERANGE) {
+            /*
+             * Overflow: strtod() returns +/- HUGE_VAL and sets errno to
+             *           ERANGE
+             *
+             * Underflow:
+             * - ANSI: strtod() returns 0 and sets errno to ERANGE
+             * - C99:  strtod() returns "a value whose magnitude is no
+             *         greater than the smallest normalized positive"
+             *         double. Whether ERANGE is set is implementation-
+             *         defined.
+             */
+            rc = errno;
+            crm_warn("Floating-point value parsed from '%s' would %sflow "
+                     "(using %g instead)", text,
+                     (fabs(local_result) > DBL_MIN) ? "over" : "under",
+                     local_result);
```

Comment:
> I think this (and later messages) should be at debug level. It's better for library functions to stick to info level or lower, if they can return a sufficiently informational return code, for two reasons

Ack, I agree with that. I'll crm_debug() the "errors" and crm_trace() the success.

> Exceptions would be legacy functions that don't return a good enough (or any) return code

I modeled these err/warn messages after scan_ll(). Say what you will about scan_ll(), but its return codes are sufficient. Do you think we should drop to crm_debug() in scan_ll()?

---

I've also removed the "ANSI" note from the comment (next push), since we're not really making an effort at compatibility with the 30-year-old standard. The underflow note will have the C99 paragraph without "C99: ".

Reply:
> I modeled these err/warn messages after scan_ll(). Say what you will about scan_ll(), but its return codes are sufficient. Do you think we should drop to crm_debug() in scan_ll()?

scan_ll() has to keep the logs (and set errno) because it's used in (public API) crm_parse_ll(), which only returns -1 on error. We're often stuck with such "exceptions", at least until the next public API compatibility break (my rough goal is to keep at least 6 years between them, to minimize the disruption for distros).